core/src/main/java/com/sequenceiq/cloudbreak/converter/stack/cluster/ServiceEndpointCollector.java

```diff
@@ -0,0 +1,189 @@
+import com.sequenceiq.cloudbreak.domain.stack.cluster.Cluster;
+import com.sequenceiq.cloudbreak.domain.stack.cluster.gateway.ExposedServices;
+import com.sequenceiq.cloudbreak.domain.stack.cluster.gateway.Gateway;
+import com.sequenceiq.cloudbreak.domain.stack.cluster.gateway.GatewayTopology;
+import com.sequenceiq.cloudbreak.domain.stack.cluster.host.HostGroup;
+import com.sequenceiq.cloudbreak.domain.stack.instance.InstanceMetaData;
+import com.sequenceiq.cloudbreak.service.network.NetworkUtils;
+
+@Component
+public class ServiceEndpointCollector {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ServiceEndpointCollector.class);
+
+    private final BlueprintValidator blueprintValidator;
+
+    private final StackServiceComponentDescriptors stackServiceComponentDescs;
+
+    public ServiceEndpointCollector(BlueprintValidator blueprintValidator, StackServiceComponentDescriptors stackServiceComponentDescs) {
+        this.blueprintValidator = blueprintValidator;
+        this.stackServiceComponentDescs = stackServiceComponentDescs;
+    }
+
+    public Map<String, String> collectServiceUrlsForPorts(Cluster cluster, String ambariIp) {
+        Set<HostGroup> hostGroups = cluster.getHostGroups();
+        Blueprint blueprint = cluster.getBlueprint();
+
+        Map<String, String> result = new HashMap<>();
+        List<Port> ports = NetworkUtils.getAllPorts();
+        try {
+            JsonNode hostGroupsNode = blueprintValidator.getHostGroupNode(blueprint);
+            Map<String, HostGroup> hostGroupMap = blueprintValidator.createHostGroupMap(hostGroups);
+            for (JsonNode hostGroupNode : hostGroupsNode) {
+                String hostGroupName = blueprintValidator.getHostGroupName(hostGroupNode);
+                JsonNode componentsNode = blueprintValidator.getComponentsNode(hostGroupNode);
+                HostGroup actualHostgroup = hostGroupMap.get(hostGroupName);
+                final String serviceAddress;
+                if (actualHostgroup.getConstraint().getInstanceGroup() != null) {
+                    InstanceMetaData next = actualHostgroup.getConstraint().getInstanceGroup().getNotDeletedInstanceMetaDataSet().iterator().next();
+                    serviceAddress = next.getPublicIpWrapper();
+                } else {
+                    serviceAddress = actualHostgroup.getHostMetadata().iterator().next().getHostName();
+                }
+                for (JsonNode componentNode : componentsNode) {
+                    String componentName = componentNode.get("name").asText();
+                    StackServiceComponentDescriptor componentDescriptor = stackServiceComponentDescs.get(componentName);
+                    Map<String, String> collectedPorts = collectServicePorts(ports, ambariIp, serviceAddress, componentDescriptor, cluster);
+                    result.putAll(collectedPorts);
+                }
+            }
+        } catch (Exception ignored) {
+            return result;
```

Comment:
I have bad feeling about not collecting something because an error, and continue the execution

Reply:
Me too but I did not want to change this part, I'm happy it works.